File: src/cpp/midi_prog/parsing/commandparser.cpp
#include "commandparser.h"
#include <regex>
#include "utils.h"
#include <sstream>
#include <fstream>
#include <iostream>

void Command::addParameter(std::string k, std::string v)
{
    this->values.push_back(v);
    this->parameters.insert(std::make_pair<std::string, std::string>(k.c_str(),v.c_str()));
}

std::string Command::getParameter(std::string key)
{
    return this->parameters[key];
}

std::string Command::getParameter(size_t index)
{
    return this->values[index];
}

bool Command::hasParameter(std::string p)
{
    return map_has_key<std::string, std::string>(this->parameters,p);
}

bool Command::hasNumberedParameter(std::string p, size_t index)
{
    p.append("_");
    p.append(std::to_string(index));
    return map_has_key<std::string, std::string>(this->parameters,p);
}

std::string Command::getNumberedParameter(std::string p, size_t index)
{
    p.append("_");
    p.append(std::to_string(index));
    return hasParameter(p) ? this->parameters[p] :  "";
}

std::string Command::getName()
{
    return this->name;
}

void Command::setName(std::string name)
{
    this->name = name;
}

std::map<std::string, std::string> Command::getParameters()
{
    return  this->parameters;
}

CommandParser::CommandParser()
{

}

std::vector<std::string> CommandParser::get_commands_file(std::string file)
{
    std::ifstream t(file);
    std::stringstream buffer;
    buffer << t.rdbuf();
    if(buffer.str() == "") std::cout << "File " << file << " couldn't be read" << std::endl;
    return CommandParser::get_commands(buffer.str());
}

std::vector<std::string> CommandParser::get_commands(std::string content)
{
    std::string current_command = "";
    std::vector<std::string> commands;
    std::vector<std::string> lines =  split(content, '\n', true);

    for(std::string line : lines)
    {
        if(line != "")
        {
            if (line[0] == '-')
            {
                current_command += line;
            }
            else if(current_command == "")
            {
                current_command = line;
            }
            else
            {
                commands.push_back(current_command);
                current_command = line;
            }
        }
    }

    if(current_command != "") commands.push_back(current_command);

    return commands;
}

std::vector<Command> CommandParser::parse_commands_file(std::string file)
{
    return CommandParser::parse_commands(CommandParser::get_commands_file(file));
}

std::vector<Command> CommandParser::parse_commands(std::vector<std::string> content)
{
    std::vector<Command> commands;

    for(std::string c : content)
    {
        commands.push_back(parse_command(c));
    }

    return commands;
}

Command CommandParser::parse_command(std::string content)
{
    Command command;
    std::regex r_params("-(\\S+) \"([^\"]+)\"");
    std::regex r_comm("([^\"\\s]+)(\\s?+)-");

    std::smatch match;

    if(std::regex_search(content, match, r_comm))
    {
        command.setName(match[1]);
    }

    std::string s = content;
    std::unordered_map<std::string, size_t> count;

    while(std::regex_search(s, match, r_params))
    {
        bool is_numbered = (match[1].str()[0] == '@');
        std::string p = static_cast<std::string>(match[1].str());
        if(is_numbered)
        {
            if(!umap_has_key<std::string, size_t>(count, p))
            {
                count.insert(std::make_pair<std::string, size_t>(p.c_str(), 0));
            }
            size_t c = count[p];
            count[p] = count[p]+1;
            p.erase(0,1);
            p.append("_");
            p.append(std::to_string(c));
        }

        command.addParameter(p, match[2]);
        s = match.suffix();
    }

    return command;
}

File: src/rust/src/lib/utils/command_parser.rs
extern crate regex;
use std::fs::File;
use std::io::prelude::*;

use std::collections::HashMap;
use regex::Regex;

pub struct Command {
    pub name: String,
    pub parameters: HashMap<String, String>,
    pub values: Vec<String>
}

impl Command {
    pub fn new() -> Command {
        Command {
            name: String::from(""),
            parameters: HashMap::new(),
            values: Vec::new()
        }
    }

    pub fn add_parameter(&mut self, k: String, v: String){
        self.values.push(v.clone());
        self.parameters.insert(k, v);
    }

    pub fn get_parameter_from_key(&self, key: &str) -> &String {
        self.parameters.get(key).expect("Key not found")
    }

    pub fn get_parameter_from_index(&self, index: usize) -> &String {
        self.values.get(index).expect("Index out of bounds")
    }

    pub fn has_parameter(&self, key: &str) -> bool {
        self.parameters.contains_key(key)
    }

    pub fn has_numbered_parameter(&self, p: &str, index: usize) -> bool {
        let mut key = String::from(p);
        key.push('_');
        key.push_str(index.to_string().as_str());
        self.parameters.contains_key(&key)
    }

    pub fn get_numbered_parameter(&self, p: &str, index: usize) -> &String {
        let mut key = String::from(p);
        key.push('_');
        key.push_str(index.to_string().as_str());
        self.get_parameter_from_key(&key)
    }
}

pub fn get_commands_file(path: &str) -> Vec<String> {
    let mut file = File::open(path).expect("Could not open file");
    let mut contents = String::new();
    file.read_to_string(&mut contents).expect("Could not read file");
    get_commands(&contents)
}

pub fn get_commands(data: &String) -> Vec<String> {
    let mut current_command: String = String::from("");
    let mut commands: Vec<String> = Vec::new();
    
    let lines = data.split('\n');

    for line in lines {
        let l: String = String::from(line.trim());
        
        if l != "" {
            if l.starts_with('-') {
                current_command += line;
            } else if current_command == "" {
                current_command = l;
            } else {
                commands.push(current_command);
                current_command = l;
            }
        }
    }

    if current_command != "" {
        commands.push(current_command);
    }

    commands
}

pub fn parse_commands_file(path: &str) -> Vec<Command> {
    parse_commands(get_commands_file(path))
}

pub fn parse_commands(data: Vec<String>) -> Vec<Command> {
    let mut commands: Vec<Command> = Vec::new();

    for c in data {
        commands.push(parse_command(&c));
    }

    commands
}

pub fn parse_command(data: &String) -> Command {
    let mut command: Command = Command::new();

    lazy_static! {
        static ref r_params: Regex = Regex::new("-(\\S+) \"([^\"]+)\"").unwrap();
        static ref r_comm: Regex = Regex::new("([^\"\\s]+)(\\s?+)-").unwrap();
    }

    let mut count: HashMap<String, usize> = HashMap::new();

    if r_comm.is_match(data) {
        let caps = r_comm.captures(data).expect("No captures found");
        command.name = String::from(caps.get(1).expect("Capture group not found for the command name").as_str());
    }

    for caps in r_params.captures_iter(data) {
        let mut p = String::from(caps.get(1).expect("Capture group not found for the command name").as_str());
        if p.starts_with("@") {
            if !count.contains_key(&p) {
                count.insert(p.clone(), 0);
            }
            let c: usize = count[&p.clone()];
            *count.get_mut(&p.clone()).expect("...") += 1;
            p.remove(0);
            p.push('_');
            p.push_str(c.to_string().as_str());
        }

        command.add_parameter(p, String::from(caps.get(2).expect("Capture group not found for the command name").as_str()));
    }

    command
}
